### ISSUE TITLE
Issue with pad_token == eos_token : model not "learning when to stop"

### ISSUE THREAD

**Issue by antmaiorino** [mlabonne/llm-course#33]
Hey @mlabonne thanks a lot for the great resources!

I have been reading the [Fine_tune_Llama_2_in_Google_Colab.ipynb](https://github.com/mlabonne/llm-course/blob/main/Fine_tune_Llama_2_in_Google_Colab.ipynb) notebook and I am encountering an issue.
I see that you have added the following lines to load the tokenizer:
```
tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right" # Fix weird overflow issue with fp16 training
```

Just to play around I have tried adapting your notebook to fine-tune a model to perform PII masking using [this dataset](https://huggingface.co/datasets/ai4privacy/pii-masking-200k) (to do it very quickly I adapted the format such that examples look like this: `<s>[INST] Mise à jour : l'heure de début de la thérapie physique a été modifiée à 8:46 AM. Lieu : Suite 348 Iva Junctions. Veuillez nous excuser pour le désagrément. [/INST] Mise à jour : l'heure de début de la thérapie physique a été modifiée à [TIME_1]. Lieu : [SECONDARYADDRESS_1] [STREET_1]. Veuillez nous excuser pour le désagrément. </s>`).

After fine-tuning the model I noticed that it was continuously generating text, effectively never producing the EOS_TOKEN and thus only stopping at the max sequence length.
By looking online it seems that this might be related to the fact that the pad token is the same as the eos token, and during training these might get masked out and thus the model might not be "learning when to stop".

Do you know if this might be the case / do you have an idea for a fix? I tried to comment out the line where you set the 2 tokens to be the same, but then my model trains for a while and then the loss suddendly drops to 0, which makes me think that that was actually needed!

**Comment by antmaiorino** [comment]
Update on the issue above:
I have been digging into the issue a bit more, and since I think I managed to solve it I'm going to add a comment in case it's useful for someone else.

Basically I noticed that the line `tokenizer.pad_token = tokenizer.eos_token` was not really useful for this model since the pad_token is already set to `<unk>` for the llama2-7b models.
Moreover, when using the `SFTTrainer` class the tokenization process automatically adds the special tokens (i.e. the BOS token `<s>` at the start of the sentence). This was also redundant as the preprocessed dataset already has a `<s>` token at the beginning, so I modified my dataset to make it start directly with `[INST]` instead.

These 2 "fixes" made the fine-tuning process work, as opposed to what happened by applying either one of these 2 or none, which was either causing the model to generate endless text or bringing the loss to 0.0 after just a couple of batches.
If anybody has more insight into why this might have been happening I'd be glad to hear it out, but in the meantime I'd say it works fine now!
